File: 264.ugly-number-ii.py
#
# @lc app=leetcode id=264 lang=python
#
# [264] Ugly Number II
#

# @lc code=start
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        import heapq
        heap = [1]
        idx = 0
        while idx < n:
            x = heapq.heappop(heap)
            while heap and heap[0] == x:
                heapq.heappop(heap)
            heapq.heappush(heap, 2 * x)
            heapq.heappush(heap, 3 * x)
            heapq.heappush(heap, 5 * x)
            idx += 1
        return x

# @lc code=end

File: 264.ugly-number-ii.cpp
/*
 * @lc app=leetcode id=264 lang=cpp
 *
 * [264] Ugly Number II
 */

// @lc code=start
class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> ret(n+1, 0);
        ret[1] = 1;
        int p2=1,  p3=1, p5 = 1;
        int num = 0;
        for(size_t i = 2; i < ret.size(); ++i){
            ret[i] = min(ret[p2] * 2, min(ret[p3] * 3, ret[p5] * 5));
            if(ret[i] == ret[p2] * 2){
                p2 += 1;
            }
            if (ret[i] == ret[p3] * 3)
            {
                p3 += 1;
            }
            if (ret[i] == ret[p5] * 5){
                p5 += 1;
            }   
        }
        return ret[n];
    }
};

// @lc code=end

